src/game.cpp

```diff
@@ -13210,11 +13235,11 @@ void game::shift_monsters( const int shiftx, const int shifty, const int shiftz
         if( shiftx != 0 || shifty != 0 ) {
             critter.shift( shiftx, shifty );
         }
-        // TODO: with z-levels, this can be removed, instead shift the critter
-        // along shiftz, too. Than make the 3D-inbounds check.
-        if( shiftz == 0 && m.inbounds( critter.posx(), critter.posy() ) ) {
+
+        if( shiftz != 0 && m.has_zlevels() && m.inbounds( critter.pos() ) ) {
```

Comment:
Should I clean it up in this PR or is it fine if I do it in next? I have active items PR almost ready.

As for the update/upgrade: upgrade should be the polymorph, but it would be good to have an update function that only makes the monster catch up in time. When monsters start regenerating etc., it would be useful to have them remember the time.